### PR TITLE
Pypardiso default

### PR DESCRIPTION
**Pull request by IvarStefansson**
## Proposed changes

The default linear solver in the models is now PyPardiso's sparse solver. 
The methods _initialize_linear_solver and assemble_and_solve were moved from ContactMechanics to AbstractModel (which is more of a base than abstract class).

## Types of changes

What types of changes does your code introduce to PorePy?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Testing (contribution related mainly to testing of existing functionality)
- [ ] Documentation update (if none of the other choices apply)
- [x] Other: Change default

## Checklist

_Enter one &check; (`&check;`) in each row. The table can be updated after creating the PR._

||Yes|No|NA|
|---|---|---|---|
|The update is covered by the  test suite |&check;|_|_|
|The documentation is up-to-date |&check;|_|_|
|I have not duplicated existing functionality |&check;|_|_|

### CODE REVIEW

File: src/porepy/models/abstract_model.py [hunk 1]
```diff
@@ -201,28 +204,89 @@ def check_convergence(
...
+            A, b = self.assembler.assemble_matrix_rhs()  # type: ignore
+        t_1 = time.time()
+        logger.debug(f"Max element in A {np.max(np.abs(A)):.2e}")
+        logger.debug(
+            f"""Max {np.max(np.sum(np.abs(A), axis=1)):.2e} and min
+            {np.min(np.sum(np.abs(A), axis=1)):.2e} A sum."""
+        )
+
+        solver = self.params["linear_solver"]
+        if solver == "pypardiso":
```
Comment (keileg) [inline src/porepy/models/abstract_model.py:263]:
```suggestion
        if solver == "pypardiso":
            # This is the default option which is invoked unless explicitly overriden by the
            # user. We need to check if the pypardiso package is available.
```

File: src/porepy/models/abstract_model.py [hunk 2]
```diff
@@ -201,28 +204,89 @@ def check_convergence(
...
+            try:
+                from pypardiso import spsolve as sparse_solver  # type: ignore
+            except ImportError:
+                sparse_solver = sps.linalg.spsolve
```
Comment (keileg) [inline src/porepy/models/abstract_model.py:267]:
```suggestion
                # Fall back on the standard scipy sparse solver.
                sparse_solver = sps.linalg.spsolve
```

**Review by keileg — APPROVED** [review]
I added some documentation. The usual caveats about line length when editing in GH reviews applies.

**Comment by IvarStefansson** [comment]
> I added some documentation. The usual caveats about line length when editing in GH reviews applies.

You nailed the line lengths! Only problem was whitespaces (trailing and in blank line)

**Pull request merged**